### DISCUSSION TITLE
Ensemble

### DISCUSSION THREAD

**Discussion by pawarbi** [microsoft/FLAML#25] [Q&A]
I ran the demo example from here: https://github.com/microsoft/FLAML/blob/main/notebook/flaml_demo.ipynb
 Couple of things: 
- I had to install openml. It would be good if flaml installed it. Should openml be added to https://github.com/microsoft/FLAML/blob/main/setup.py considering that's the first thing required to run the demo datasets using   `from flaml.data import load_openml_dataset`
- Does setting `ensemble=True` in `settings` force flaml to create and output an ensemble model? Expected behavior should be if  `ensemble=True`, train individual models and ensembles but return the best model based for the chosen metric. If you run the demo notebook from above, an ensemble model is returned which has lower accuracy than the single best model (lgbm). Expected output in this case would be best LGBMClassifier model not the StackingClassifer which has lower accuracy.

**Comment by sonichi** [comment]
1. For openml, we made it optional in the default package. In the `notebook` option, openml is included. So one can use `pip install flaml[notebook]` to make sure the notebook is runnable. I'll add `!pip install flaml[notebook]` in the beginning of the notebook so that it ensures the dependencies are installed.
2. Without access to test data we won't know whether the best single model is better or the ensemble is better. What we can do is to keep both models available by adding a property `best_individual_model` when `ensemble=True`. Would you like to work on that?

**Reply by pawarbi** [reply]
1. Noted, thanks. Makes sense. 
2. Why would test set be needed? Isn't the best model from the training based on the validation set (or mean of the metric over kf)? Just like the current approach, log error metric for individual and ensemble models, return the leaderboard with model with the best performance. User will evaluate the generalized performance on the test set for the returned best model. Am I missing something.

**Reply by sonichi** [reply]
The current ensemble uses the full training data, so we can't directly compare its error with the base model. We can allow users to set `ensemble='if_better'`. When they do so, we can insert a step of building an ensemble using the same evaluation procedure as base models, and keep the ensemble only if it's better than the best base model. Feel free to create a PR about it.

**Reply by wjlgatech** [reply]
Hi @sonichi, is the `emsemble` option available now? Is that a demo for its use? Thanks

**Reply by sonichi** [reply]
ensemble – boolean or dict | default=False. Whether to perform ensemble after search. Can be a dict with keys ‘passthrough’ and ‘final_estimator’ to specify the passthrough and final_estimator in the stacker.

https://github.com/microsoft/FLAML/blob/ddc1a63a764d546e902acd3599c2de387ab2673e/test/test_automl.py#L253
https://github.com/microsoft/FLAML/blob/ddc1a63a764d546e902acd3599c2de387ab2673e/test/test_automl.py#L164

**Reply by wjlgatech** [reply]
@sonichi Thanks for the references. When you set `ensemble=True`, after `automl.fit(X_train=X_train, y_train=y_train, **settings)`, is there a place you can extract the ensemble model? and do something like this

```python
#pseudo code here:
y_pred = automl.ensemble_model.predict(X_test)
y_prob = automl.ensemble_model.predict_proba(X_test)
```

**Reply by wjlgatech** [reply]
Moreover, what are the ensemble methods FLAML use? 

Is it a simple average or more sophisticate one like linear regression/MLP on the top of all trained models?

```python
# simple average ensemble
y_prob_average = (y_prob1 + y_prob2 + y_prob3)/3
```

**Reply by sonichi** [reply]
We use StackingClassifier or StackingRegressor from sklearn.ensemble. When `ensemble=True`, you can directly use `automl.predict` or `automl.predict_proba` to apply the ensemble model. To extract the ensemble model, use `automl.model`.
